### PR TITLE
Add heroku metadata monitor

### PR DESCRIPTION
**Pull request by asuresh4**
Collects Heroku dyno metadata if enabled: https://devcenter.heroku.com/articles/dyno-metadata#dyno-metadata

### CODE REVIEW

File: docs/monitors/heroku-metadata.md [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+# heroku-metadata
+
+Monitor Type: `heroku-metadata` ([Source](https://github.com/signalfx/signalfx-agent/tree/master/pkg/monitors/heroku))
+
+**Accepts Endpoints**: **Yes**
+
+**Multiple Instances Allowed**: Yes
+
+## Overview
```
Comment (keitwb) [inline docs/monitors/heroku-metadata.md]: This will need docs on how to enable the metadata in Heroku.

File: pkg/monitors/heroku/monitor.go [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+		properties["heroku_app_name"] = os.Getenv("HEROKU_APP_NAME")
+		properties["heroku_slug_commit"] = os.Getenv("HEROKU_SLUG_COMMIT")
+		properties["heroku_release_creation_timestamp"] = os.Getenv("HEROKU_RELEASE_CREATED_AT")
+		properties["heroku_app_id"] = os.Getenv("HEROKU_APP_ID")
+
+		m.Output.SendDimensionUpdate(&types.Dimension{
+			Name:              "dyno_id",
+			Value:             dynoID,
+			Properties:        properties,
+			MergeIntoExisting: true,
```
Comment (keitwb) [inline pkg/monitors/heroku/monitor.go]: You probably want this to be `false` unless you know of something else manipulating the dyno_id dimension.

File: pkg/monitors/heroku/monitor.go [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+func init() {
+	monitors.Register(&monitorMetadata, func() interface{} { return &Monitor{} }, &Config{})
+}
+
+// Configure monitor
+func (m *Monitor) Configure(c *Config) error {
+	m.logger = utils.NewThrottledLogger(log.WithFields(log.Fields{"monitorType": "heroku-metadata"}), 20*time.Second)
+	m.ctx, m.cancel = context.WithCancel(context.Background())
+
+	utils.RunOnInterval(m.ctx, func() {
```
Comment (keitwb) [inline pkg/monitors/heroku/monitor.go]: This doesn't need to run on an interval, only once.

Comment (keitwb) [reply pkg/monitors/heroku/monitor.go]: But do the run in a goroutine.

File: pkg/monitors/heroku/monitor.go [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+			MergeIntoExisting: true,
+		})
+
+	}, time.Duration(c.IntervalSeconds)*time.Second)
+
+	return nil
+}
+
+// Shutdown stops the metric sync
+func (m *Monitor) Shutdown() {
```
Comment (keitwb) [inline pkg/monitors/heroku/monitor.go]: No need for a shutdown.

File: pkg/monitors/heroku/monitor.go (hunk 2 above)
Comment (asuresh4) [reply pkg/monitors/heroku/monitor.go]: Removed

File: pkg/monitors/heroku/monitor.go (hunk 3 above)
Comment (asuresh4) [reply pkg/monitors/heroku/monitor.go]: Updated

**Comment by keitwb** [comment]
So the stuff from the collector app that will have dyno_id is the memory/load metrics and any custom metrics emitted by the app?  But router metrics will only have dyno name as a dimension?

File: pkg/monitors/heroku/monitor.go [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+
+	go func() {
+		properties := map[string]string{}
+		dynoID := os.Getenv("HEROKU_DYNO_ID")
+
+		properties["heroku_release_version"] = os.Getenv("HEROKU_RELEASE_VERSION")
+		properties["heroku_app_name"] = os.Getenv("HEROKU_APP_NAME")
+		properties["heroku_slug_commit"] = os.Getenv("HEROKU_SLUG_COMMIT")
+		properties["heroku_release_creation_timestamp"] = os.Getenv("HEROKU_RELEASE_CREATED_AT")
+		properties["heroku_app_id"] = os.Getenv("HEROKU_APP_ID")
```
Comment (keitwb) [inline pkg/monitors/heroku/monitor.go]: Is there any way to get the dyno type?  e.g. Hobby, etc.

**Comment by asuresh4** [comment]
> So the stuff from the collector app that will have dyno_id is the memory/load metrics and any custom metrics emitted by the app? But router metrics will only have dyno name as a dimension?

yes, that's right

File: pkg/monitors/heroku/monitor.go (hunk 5 above)
Comment (asuresh4) [reply pkg/monitors/heroku/monitor.go]: I looked through the docs, I don't this information being exposed through the dyno metadata "endpoint". Seems like this information is only exposed via the platform api.

File: docs/monitors/heroku-metadata.md (hunk 1 above)
Comment (asuresh4) [reply docs/monitors/heroku-metadata.md]: Added docs

**Comment by asuresh4** [comment]
> > So the stuff from the collector app that will have dyno_id is the memory/load metrics and any custom metrics emitted by the app? But router metrics will only have dyno name as a dimension?
> 
> yes, that's right

This not actually true. Custom metrics being logged by application will does not have `dyno_id`. Here's an example,

`164 <190>1 2019-12-21T22:21:26.705132+00:00 host app web.1 - gauge#quota_used=20 cumulative#response_bytes=100 sfxdimension#service=backend sfxdimension#client=sfx_app`

where the message

`gauge#quota_used=20 cumulative#response_bytes=100 sfxdimension#service=backend sfxdimension#client=sfx_app`

is what is logged out from an application and it will not contain Dyno id. However, it could be added by users since the application has access to the env variables on the dyno.

**Pull request merged**